src/main/java/org/mqttbee/mqtt/handler/publish/MqttIncomingQosHandler.java

```diff
@@ -118,13 +116,20 @@ private void handlePublishQos1(
 
         final Object previousMessage = messages.put(publish.getPacketIdentifier(), MqttPubAck.class);
         if (previousMessage == null) { // new message
-            emitOrQueuePublishQos1Or2(ctx, publish);
-        } else if (previousMessage == MqttPubAck.class) { // resent message (MQTT 3)
+            if (!getIncomingPublishService().onPublish(publish)) {
+                disconnectReceiveMaximumExceeded(ctx);
+            }
+        } else if (previousMessage == MqttPubAck.class) { // resent message
             if (!publish.isDup()) {
-                disconnectDupFlagNotSet();
+                disconnectDupFlagNotSet(ctx);
             }
-        } else { // packet id in use
-            ackQos1(ctx, new MqttPubAckBuilder(publish).reasonCode(Mqtt5PubAckReasonCode.PACKET_IDENTIFIER_IN_USE));
+        } else if ((previousMessage == MqttPubRec.class) || (previousMessage == MqttPubComp.class)) { //packet id in use
+            //ackQos1(ctx, new MqttPubAckBuilder(publish).reasonCode(Mqtt5PubAckReasonCode.PACKET_IDENTIFIER_IN_USE));
+            MqttDisconnectUtil.disconnect(ctx.channel(), Mqtt5DisconnectReasonCode.PROTOCOL_ERROR,
+                    "Packet Identifier in use: QoS 1 Publish must not be received with the same Id as a QoS 2 Publish");
+        } else { // PubRel: packet id in use
+            MqttDisconnectUtil.disconnect(ctx.channel(), Mqtt5DisconnectReasonCode.PROTOCOL_ERROR,
+                    "Packet Identifier in use: QoS 1 Publish must not be received with the same Id as a PubRel");
```

Comment:
Try to avoid code duplication:
```
else {
    String disconnectReason = ...
    if (...) {
         disconnectReason = ...
    } else {
         disconnectReason = ...
   }
   disconnect(...)
```